Read BGC input files through pkg/gchem

averdy opened the pull request:
## What changes does this PR introduce?
Feature:  BGC input files are handled through pkg/gchem. A more general approach than what was initially proposed for pkg/bling ( PR #841 ), as it can be used for other BGC models. See issue #950 and issue #138 

## What is the current behaviour? 
BGC forcing (surface iron flux, atmospheric pCO2, wind speed, atmospheric pressure, ice) and fields that are not prognostic variables (surface silica) are specified in namelist files data.dic, data.bling, or data.darwin and read in the package-specific *_fields_load.F routine. 

## What is the new behaviour 
Those fields are now read in gchem_fields_load.F and handled through data.gchem. It is done using pkg/exf routines if useEXF (or maybe useGCHEM_EXF?) is true; otherwise it is done using either the current method (READ_REC_XY_RS).

## Does this PR introduce a breaking change? 
For pkg/bling, the only change required is switching BGC forcing fields info to data.gchem. For other packages, no breaking change - by default, forcing fields read with gchem will not be passed on to the package. 

## Other information:
- For winds and atmos pressure, need to check compatibility of exf fields (wspeed and apressure) with BGC code
- Currently doesn't run with #define USE_EXF_INTERPOLATION
- Eventually we want to read in 3D silica fields

## Suggested addition to `tag-index`

averdy committed:
input for new verif expt

averdy committed:
remove unused variables

mjlosch commented:
Hi @averdy
thanks you for making a great start here. Could you please modify this PR, so that it is easier to compare, e.g. 
- the directory `gchem` should contain all the fields necessary to compile the model (I guess it does), could you please remove `gchem/gchem_orig` (we don't need the originals here, git takes care of that).
- the new verification experiment `input.exf` contains duplicates of `input` and `input_ad`. I assume that it is a fwd experiment, so the ad-specific file are maybe not needed at all, and the rest of the duplicates should be (automatically) linked from `input`, so they can be removed, too, e.g. `data.gmredi`. It makes it easier to concentrate on the relevant changes.
- If you add an `output.exf.txt` to `results`, it would be easier to test this

I'll have a close look, once this is done, but for naming I suggest that the field names need not necessarily include "exf", because one can image reading the same fields via the "old" method. For a cleaner separation, we could think of having a separate routine `gchem_exf_fields_load` or `gchem_exf_getffields.F` that is called from (or before) `gchem_fields_load`.

jm-c commented:
@averdy Thanks for starting this. I suggest to wait a little bit before any new updates here to give us some time to provide feedback.

averdy committed:
Delete pkg/gchem/gchem_orig directory

averdy committed:
remove duplicate input files

mjlosch commented:
I needed to have a data with
```
#tauThetaClimRelax = 5184000.,
#tauSaltClimRelax  = 7776000.,
``` 
commented out, but with these changes:
```
diff --git a/pkg/gchem/gchem_init_vari.F b/pkg/gchem/gchem_init_vari.F
index a877b1c53..d5926e406 100644
--- a/pkg/gchem/gchem_init_vari.F
+++ b/pkg/gchem/gchem_init_vari.F
@@ -169,8 +169,8 @@ C-- Initialize Geo-Chemistry forcing fields
         _EXCH_XY_RL( gchemSi, myThid )
         _EXCH_XY_RL( gchemPAR, myThid )
         _EXCH_XY_RL( gchemFe, myThid )
-        _EXCH_XY_RL( gchemIce, myThid )
-        _EXCH_XY_RL( gchemWind, myThid )
+c       _EXCH_XY_RL( gchemIce, myThid )
+c       _EXCH_XY_RL( gchemWind, myThid )
         _EXCH_XY_RL( gchemapCO2, myThid )
 
 #endif /* ALLOW_EXF */
diff --git a/pkg/gchem/gchem_readparms.F b/pkg/gchem/gchem_readparms.F
index 8968149f4..bff3ce516 100644
--- a/pkg/gchem/gchem_readparms.F
+++ b/pkg/gchem/gchem_readparms.F
@@ -102,7 +102,7 @@ C- Open and read the data.gchem file
      I                   myThid )
       READ(UNIT=iUnit,NML=GCHEM_PARM01)
 #ifdef ALLOW_EXF
-      CALL GCHEM_EXF_READPARMS( iUnit, myThid )
+      IF (useEXF) CALL GCHEM_EXF_READPARMS( iUnit, myThid )
 #endif /* ALLOW_EXF */
       WRITE(msgBuf,'(A)')
      &  ' GCHEM_READPARMS: finished reading data.gchem'
diff --git a/verification/global_oce_biogeo_bling/input.exf/data.pkg b/verification/global_oce_biogeo_bling/input.exf/data.pkg
index 2564a8df7..04701bb30 100644
--- a/verification/global_oce_biogeo_bling/input.exf/data.pkg
+++ b/verification/global_oce_biogeo_bling/input.exf/data.pkg
@@ -3,7 +3,7 @@
  useCAL=.TRUE.,
  useEXF=.TRUE.,
  useGMRedi=.TRUE.,
- usePROFILES=.TRUE.,
+#usePROFILES=.TRUE.,
  usePTRACERS=.TRUE.,
  useGCHEM=.TRUE.,
  useDiagnostics=.TRUE.,
```
I can now run input.exf, but the default fwd experiment fails. 
@averdy Did you try that?

averdy commented:
@mjlosch I'm having computing issues and can't run the model right now, but I'll look into it as soon as possible! I'm sharing this as a draft mostly to get feedback on the general structure. I'm not attached to any of it at this point!

averdy committed:
Fix some bugs

averdy committed:
One more bug

mjlosch committed:
only initialise exf files when useEXF=T

avoids using unitialised filenames that make reference experiment
global_oce_biogeo_bling fail (and it is better style)

mjlosch commented on pkg/bling/bling_fields_load.F:
```diff
@@ -35,6 +27,12 @@ SUBROUTINE BLING_FIELDS_LOAD (
 #endif
+#ifdef ALLOW_EXF
+# include "GCHEM_EXF.h"
+# include "EXF_PARAM.h"
+# include "EXF_INTERP_SIZE.h"
+# include "EXF_FIELDS.h"
```
are these needed here? Everything should be `GCHEM_EXF.h` right?

mjlosch replied:
I guess, they are needed because of the actual exf fields being used here. I think `EXF_INTERP_SIZE.h` can be removed, though.

averdy replied:
These are removed in the latest version, since everything is in GCHEM_EXF.h now

mjlosch commented on pkg/bling/bling_fields_load.F:
```diff
@@ -270,25 +268,31 @@ SUBROUTINE BLING_FIELDS_LOAD (
 C-----------------------------------------------------------
-C Get Atmospheric carbon dioxide concentration from PKG/EXF
+C Get EXF fields from gchem
+      DO bj = myByLo(myThid), myByHi(myThid)
+       DO bi = myBxLo(myThid), myBxHi(myThid)
```
could be inside `IF ( useExF ) THEN`/`ENDIF`

mjlosch replied:
To me it would make sense to have one big
```
#ifdef ALLOW_EXF
      IF ( useEXF ) THEN
       DO bj = myByLo(myThid), myByHi(myThid)
        DO bi = myBxLo(myThid), myBxHi(myThid)
         [...]
        ENDDO
       ENDDO
      ENDIF
#endif
```
that includes all the exf-related code, and maybe even exclude the non-exf code by an `ELSE`.

Then again, the non-exf code could also be moved to `gchem`, because it is also not very different between `bling` and `dic` and then one could make this routine entirely agnostic of `pkg/exf` and it just gets its fields from `pkg/gchem`.

averdy replied:
This was moved out of bling_fields_load.F

mjlosch commented on pkg/bling/bling_light.F:
```diff
@@ -155,58 +160,63 @@ SUBROUTINE BLING_LIGHT(
+        STOP 'ABNORMAL END: S/R INSOL: 2-D output not implemented'
+       ENDIF
+
+C Find day (****NOTE for year starting in winter*****)
```
This goes a little further, but this bit is basically a copy of `dic/insol.F`, isn't it? Would it make sense to have a routine `pkg/gchem/gchem_insol.F` (or a similar name) that then handles the computation of `irr_surf` if `irr_surf` is not read from file? 
Or even have this routine `pkg/gchem/gchem_insol.F` return some `irr_surf` that is either computed from these astronomical parameters or retrieved from a file, either by exf or not, it could even include the `QSW_underice` part.

mjlosch reviewed (commented):
I think that this is great. I have a few inline comments and suggestions.

I would even go a step further and make `pkg/bling` (and `pkg/dic` in a similar way) entirely agnostic of where the forcing fields come from. A `gchem_fields_load` loads all all forcing data provided by `data.gchem` into fields called `gchem_wspeed`, etc, and it is here, where we decide if we want to use the ext-method or the original simple method. `bling_fields_load` would then just copy to forcing fields local to `bling` (or `dic`), or uses the gchem-fields explicitly (i.e. `bling_fields_load` would no longer be necessary). Each BGC package called from `gchem` can use the fields or decide not to use them and have their own method of loading them.

mjlosch commented on pkg/gchem/gchem_fields_load.F:
```diff
@@ -50,10 +60,78 @@ SUBROUTINE GCHEM_FIELDS_LOAD (
+     &     PAR_nlon, PAR_nlat, xC, yC,
+     &     PAR_interpMethod,
+#endif
+     &     mytime, myiter, mythid )
+       ENDIF
+
+       IF ( exf_ironfile .NE. ' '  ) THEN
+       CALL EXF_SET_FLD(
+     &     'iron', exf_ironfile, ironmask,
+     &     ironStartTime, ironperiod, ironRepCycle,
+     &     gchem_inscal_iron,
+     &     iron_exfremo_intercept, iron_exfremo_slope,
+     &     gchemFe, giron0, giron1,
+#ifdef USE_EXF_INTERPOLATION
+     &     iron_lon0, iron_lon_inc,
+     &     iron_lat0, iron_lat_inc,
+     &     iron_nlon, iron_nlat, xC, yC,
+     &     iron_interpMethod,
+#endif
+     &     mytime, myiter, mythid )
+       ENDIF
+
+       IF ( exf_apco2file .NE. ' '  ) THEN
+       CALL EXF_SET_FLD(
+     &     'apCO2', exf_apCO2file, apCO2mask,
+     &     apCO2StartTime, apCO2period, apCO2RepCycle,
+     &     gchem_inscal_apCO2,
+     &     apCO2_exfremo_intercept, apCO2_exfremo_slope,
+     &     gchemapCO2, apCO20, apCO21,
+#ifdef USE_EXF_INTERPOLATION
+     &     apCO2_lon0, apCO2_lon_inc,
+     &     apCO2_lat0, apCO2_lat_inc,
+     &     apCO2_nlon, apCO2_nlat, xC, yC,
+     &     apCO2_interpMethod,
+#endif
+     &     mytime, myiter, mythid )
+       ENDIF
+      ENDIF
+      
+#endif /* ALLOW_EXF */
```
this should be before `bling_fields_load`, etc. otherwise these fields are not available in time

averdy replied:
bling_fields_load.F is now called from gchem_fields_load.Fm after the lines highlighted above

mjlosch commented:
> I would even go a step further and make pkg/bling (and pkg/dic in a similar way) entirely agnostic of where the forcing fields come from. 

I have a version in my local repo, where I have done this. It gives exactly the same solution (via testreport) as the current code. 
To show this to you without modifying this branch I am posting a patch that you can apply to your repo by
```
git apply gchem-bling-exf.patch
```
[gchem-bling-exf.patch](https://github.com/user-attachments/files/17395816/gchem-bling-exf.patch)

It's not complete (e.g. some of the runtime parameters such as `BLING_pCO2` need to be transferred to `gchem`, etc. but the main structure is what I meant in my earlier description. If we were to go in this direction, one would need to sort out the namelists (no clear separation between filenames with and without exf), etc.

seamanticscience commented on pkg/gchem/GCHEM_EXF.h:
```diff
@@ -0,0 +1,268 @@
+      _RL  ice_lat_inc(MAX_LAT_INC)
+#endif
+
+C Wind forcing parameters for exf
+
+      _RL windStartTime
+
+      COMMON/gchem_forcing_wind_c/
+     &    windmask
+      COMMON/gchem_forcing_wind_i/
+     &    windstartdate1, windstartdate2
+      COMMON/gchem_forcing_wind_r/
+     &    windStartTime,
+     &    windperiod, windRepCycle, windconst,
+     &    wind_exfremo_intercept, wind_exfremo_slope,
+     &    gchem_inscal_wind
+      CHARACTER*1 windmask
+      INTEGER windstartdate1
+      INTEGER windstartdate2
+      _RL windperiod
+      _RL windRepCycle
+      _RL windconst
+      _RL wind_exfremo_intercept
+      _RL wind_exfremo_slope
+      _RL gchem_inscal_wind
+
+#ifdef USE_EXF_INTERPOLATION
+      COMMON/gchem_interp_wind_i/
+     &    wind_nlon, wind_nlat, wind_interpMethod
+      COMMON/gchem_interp_wind_r/
+     &    wind_lon0, wind_lat0, wind_lon_inc,
+     &    wind_lat_inc
+      INTEGER wind_interpMethod, wind_nlon, wind_nlat
+      _RL  wind_lon0
+      _RL  wind_lat0
+      _RL  wind_lon_inc
+      _RL  wind_lat_inc(MAX_LAT_INC)
+#endif
+
```
This is for wind speed, right? `pkg/exf` already has machinery for loading wind speed (`wspeed`) or calculating it from wind stress in `S/R exf_wind.F`.

averdy replied:
Thanks @seamanticscience  I removed it, as well as the ice-related variables, since neither is needed in gchem (I previously had them in gchem_exf_readparms.F but not anymore)

seamanticscience replied:
Hi @averdy, just to keep track, `fice` (bling) <- `areamask` (exf) and `wind` (bling) <- `wspeed` (exf)? Wondering if there should be a warning if wind speed is to be read in, but exf option `useAtmWind=TRUE`?

averdy committed:
Removing wind and ice related variables

mjlosch committed:
Revert "Removing wind and ice related variables"

This reverts commit eb83575e9ad7f360681631d4a5b6c2c13e0fd0e7.

mjlosch committed:
move all forcing i/o from pkg/bling to pkg/gchem

- bling_fields_load.F only contains a copy and possibly a unit
  conversion
- replace exf_ prefix by gchem_ (not fully consistent yet)
- always call gchem_exf_readparams
- gchem_fields_load.F contains all reading of forcing files, with and
  without exf; logic may not yet be optimal when to read with pkg/exf
- some sensible initialisation of fields in gchem_init_fixed.F is
  necessary to main results of reference experiment
- adjust namelists of reference experiment(s)
- add preliminary reference output for testing
- this does not yet work with TAF AD because variable names have
  changed and need adjustment; postponed to when naming changes are complete

mjlosch committed:
quick fixes to fix some verification experiments with gchem

tutorial_cfc_example and tutorial_global_oce_biogeo now compile and run

mjlosch commented:
I did not pay to much attention to other verification experiments than `global_oce_biogeo_bling`, so that there were some problem, when `bling` was not used. The fixes in 82ea1c5 addressed this, but we may want to think about a different organisation of `gchem_readparms.F` and `gchem_exf_readparms.F`.

I note that none of this works with TAF, because variable names have changed and variable have moved. We will sort this out later, when we have converged on a complete set of names of the new gchem forcing variables.

A few thoughts:
- We could use `gchem_` or `gc_` to prepend variables where necessary. I prefer `gchem_` (pkg-name). 
- Many of the new variables will only be used in `gchem_fields_load.F`, especially `gSI0/1` etc., so we may want keep them in a separate header file from the actual forcing field `gchemSi` (or maybe `gchem_Si`).
- The name `GCHEM_EXF.h` may also not be a good choice, because these fields are also used if `pkg/exf` is not used. 
- I suggest to move the forcing fields `gchemSi` (and all others) to `GCHEM.h` or a new file `GCHEM_FORCING.h` and rename `GCHEM_EXF.h` to `GCHEM_FIELDS.F` or similar for the remaining `gSi0/1` etc.

mjlosch committed:
move 2D forcing fields to GCHEM_FIELDS.h, rearrange namelists

- rename forcing fields to have prefix "gchem" (for now without _)
- forcing file names in general gchem namelist, maybe it is even
  better to move all of gchem_exf_readparms.F to gchem_readparms.F,
  including the forcing namelist?
- foward testreport passes (at least on my computer)

averdy committed:
remove fields loading in bling

averdy committed:
Add small missing pieces (initializations, warnings)

averdy committed:
Read winds and atmosP with pkg/exf

mjlosch commented:
@averdy the latest commit breaks some of the verification experiments with `gchem` that do now use `pkg/exf` (and don't compile it either). Maybe some `#ifdef ALLOW_EXF` are missing?

averdy committed:
Add missing ALLOW_EXF

averdy commented:
@mjlosch Fixed, thanks!

averdy committed:
Fix if/do loops

averdy committed:
Fix typos so can compile with USE_EXF_INTERPOLATION

mjlosch committed:
sort out logic and bi/bj loops: exf_set_fld must not be inside bi/bj

mjlosch commented:
@averdy not sure why, but [2495e36](https://github.com/MITgcm/MITgcm/pull/876/commits/2495e365b7e1dfe65dc7f68ecfe39fc4243a46f6) changes the results of `verification/global_oce_biogeo_bling`. Do you have any idea why? It would be good to fix this.

mjlosch commented:
The largest changes are from [a63c177](https://github.com/MITgcm/MITgcm/pull/876/commits/a63c1772d548a53c370615f378fdf7aa251752e7) where initial conditions are changed. I think we should stick to the old initial conditions to be able to reproduce the old results for now, so that we don't get confused. In a separate PR we can still change these defaults if really necessary.
The remaining changes are from [2495e36](https://github.com/MITgcm/MITgcm/pull/876/commits/2495e365b7e1dfe65dc7f68ecfe39fc4243a46f6). They are not big, but still it would be good to understand why this happens (and potentially fix it):
```
default 10  ----T-----  ----S-----  ----U-----  ----V-----  --PTR 01--  --PTR 02--  --PTR 03--  --PTR 04--  --PTR 05--
G D M    c        m  s        m  s        m  s        m  s        m  s        m  s        m  s        m  s        m  s
e p a R  g  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .
n n k u  2  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d
2 d e n  d  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .

Y Y Y Y 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16  8 16 10> 7<16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 FAIL  global_oce_biogeo_bling
Y Y Y Y 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16  8 16 10> 7<16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 FAIL  global_oce_biogeo_bling.exf
```

jahn commented:
Not sure I understand why gchem_pco2 is needed.  The exf package already has a mechanism for setting a constant value.  Why not use apCO2const?  If I understand the proposed changes, this is a runtime parameter but ignored.  This is kind of unexpected.  Same for the hard-coded default values for the other fields set in gchem_init_varia.F.  Apologies if I am missing something, just trying to catch up.

mjlosch commented:
There was `bling_pCO2`, so I thought I'd replace it. This is for the case that `pkg/exf` is not defined. 

I totally see that this is redundant, and we can use `apCO2const` instead (but then need to reorganise name lists and common blocks). For this we'd probably need to merge gchem_readparms.F and gchem_exf_readparms.F, or at least have the GCHEM_EXF.h parameters defined all the time (and included in gchem_readparms.F). 

The hard coded default values are only here to be able to reproduce old results (which already does not work anymore). This is now modelled a bit after the bling and dic-verification experiments, because that's what we have. There is no problem in modifying the defaults as soon as there's a consensus what they should be (probably not zero in some cases).

averdy commented:
@mjlosch I think the issue with the test results is that gchem_fields_load.F is not called before bling_carbonate_init.F, and silica values are needed for pCO2 calculations

mjlosch commented:
@averdy after merging `global_oce_biogeo_bling` does not compile anymore.

averdy committed:
Update apco2 file name

averdy commented:
@mjlosch The forward code now gives the same results as previously. However, it doesn't when pkg/exf is compiled - I'm looking into this now.

mjlosch commented:
@averdy I suggest to go back to [0c1d330](https://github.com/MITgcm/MITgcm/pull/876/commits/0c1d330e900e5bba7010852c210d7dfbf615b0c1) and retrace the steps figuring out the reasons for changed results in each (sub-)step. I would hold hold off merging master into this branch until the very end. I have the impression that the merge [2caab55](https://github.com/MITgcm/MITgcm/pull/876/commits/2caab55891a7888af27f4a43fe6fd09c3bfb809e) makes things much more complicated.

averdy commented:
@mjlosch Sure, that sounds great!

mjlosch commented:
I think the problem we are facing has to do with unfortunate initialisation before commit [2495e36](https://github.com/MITgcm/MITgcm/pull/876/commits/2495e365b7e1dfe65dc7f68ecfe39fc4243a46f6)
1. In the experiment without exf, field `silica` (and other fields) is set to some constants in `gchem_init_varia.F` (and originally in `bling_ini_forcing.F`), but then only the field `silica` is initialized from the time dependent file `BLING_silicaFile`. This could be copied to `gchem_init_varia.F` with `gchem_silicafile`.
2. in (new) the experiment with `exf`, the constant `gchemSi` is overwritten by the first record of `gchem_silicaFile`, but then it is overwritten again in `bling_ini_forcing.F` by a constant.

I'd argue that both initialisations make little sense. In (1), we should initialise the 2D fields by constants and then read the first record of the forcing ***for all fields***  (as is done for the `exf`-case). In (2) overwriting the loaded field by a constant is not good either. I suggest that we do the modifications before commit [2495e36](https://github.com/MITgcm/MITgcm/pull/876/commits/2495e365b7e1dfe65dc7f68ecfe39fc4243a46f6), then re-generate the reference result, and then continue with the commits starting with [2495e36](https://github.com/MITgcm/MITgcm/pull/876/commits/2495e365b7e1dfe65dc7f68ecfe39fc4243a46f6), to make sure the the following changes do not change the results again. From a git point of view, this is going to be a bit of a mess (and may have to involve `git rebase`. But I can try to do that if you agree with this procedure, @averdy. 

Alternatively we can decide that we do not care about reproducing old results and just regenerate them now, but then you'd have to check carefully (in longer runs) that the new results make sense.

averdy commented:
@mjlosch I'm ok with anything! Sorry if I made things more complicated. I'm happy to let you take the lead from here on!

mjlosch committed:
change of initialisation of bling changes results

- remove reading silica from bling_carbonate_init.F
- copy gchem-fields to bling fields in bling_ini_forcing.F
- after initialisation with constants, read initial conditions from
  file if specified (as is done in, e.g., model/src/ini_forcing.F)
- adjust namelist parameters of global_oce_biogeo_bling

mjlosch committed:
update reference results after modifying initialisation of bling

- preliminary, because generated on my laptop

mjlosch committed:
restore defaults to keep global_oce_biogeo_bling unchanged

mjlosch commented:
@averdy and everyone. I think I know what happened: 
1. some commits in this PR changed results, which are a bit hard to recover because the intialisation in bling is a little inconsistent. 
2. there was a change in defaults (specifically `#undef USE_QSW`) that greatly affects the results (updated on Nov6), so that merging the upstream/master changes everything here, too.

As a way forward I suggest this procedure:
I branched off before [2495e36](https://github.com/MITgcm/MITgcm/pull/876/commits/2495e365b7e1dfe65dc7f68ecfe39fc4243a46f6), modified the initialisations in a way that to my mind makes the most sense (and others may disagree). This changes results a little and I updated the `output.txt/output.exf.txt`. Then by cherry-picking I brought the new branch up-to date with this branch (commit [b9a24db](https://github.com/MITgcm/MITgcm/pull/876/commits/b9a24db4adeae6b330b0e4465160b5bc0db790df)) minus the merge with master ([2caab55](https://github.com/MITgcm/MITgcm/pull/876/commits/2caab55891a7888af27f4a43fe6fd09c3bfb809e)), making sure that the (fwd!) results do not change along the way.

The next step would be to force-push my new branch to this branch `gchem_exf`. This will rewrite the history and will make it necessary for you to reset your local copies of the branch.

Then, we need to decide, if we want to merge upstream/master and update the results as the next step or hold off doing that until the very end. There is always going to be a big difference between the results of this branch and upstream/master, just because the results on master have changed to much for `global_oce_biogeo_bling`.

Alternative suggestions? (@jm-c)

averdy commented:
@mjlosch Sounds good to me!

mjlosch commented:
OK, I have force-pushed my changes. @averdy and anyone who has a copy of this branch locally need to reset this copy by
```
git fetch origin
git checkout gchem_exf
git reset --hard origin/gchem_exf
```

If you are not `averdy`, then you need to replace `origin` by `averdy` or whatever name you have given to the repository `github.com:averdy/MITgcm.git`

Alternatively you can delete the branch locally (`git branch -d gchem_exf`) and make a new checkout.

mjlosch committed:
merge gchem_exf_readparms.F with gchem_readparms.F, universal const.

prepend constants with "gchem_", e.g. gchem_par_const, etc. and give
the default values other than UNSET_RL, except of gchem_silica_const,
because that changes results (again)

mjlosch committed:
give gchem_silica_const a reasonalbe default and update output.txt

mjlosch committed:
some cleanup, GCHEM_EXF.h is only included within ifdef ALLOW_EXF

mjlosch commented:
I included all of `gchem_exf_readparms.F` into `gchem_readparms.F` and reoganized the `gchem_readparms.F` namelists and gchem common blocks so that we use the same constants for initialising gchem forcing fields regardless of whether `pkg/exf` is used or not (this required another small change to `output.exf.txt`. I don't like that both file names and constants are part of namelist `GCHEM_PARM01` and the rest is part of `GCHEM_FORCING_PARAMS`, but for backward compatibility reasons this was necessary.

One could do more cleanup and separate the forcing fields in `GCHEM_FIELDS.h` from the auxiliary fields (e.g. `gchemSi0/gchemSi1`) in a new file, e.g. `GCHEM_FIELDS_AUX.h`, so that this file with the auxiliary fields is only visible in `gchem_fields_load.F` and `gchem_init_vari.F`, but not outside of `pkg/gchem`.

mjlosch committed:
more cleanup: always useEXF with ALLOW_EXF and rearrangements

mjlosch committed:
separate gchem fields from auxillary fields

mjlosch committed:
remove ALLOW_GCHEM where not necessary

mjlosch commented:
@averdy, @seamanticscience If you find the time, it would be great if you could talk a look at the structure and maybe confirm in a longer run that the results are not changed (very much) for `bling`, as the initialisation has changed. After that we could try to merge in master (which will break all results but we know why and can just update the reference).

The question about units remains. Currently there's no description of forcing fields in `GCHEM_FIELDS.h`, so units are arbitary. I cannot remember what were the pros and cons of settling for specific units, but I see that
- on the one hand, it would be nice to be able to switch between models without having to change the forcing fields (settle for a unit and do the conversion in the respective model)
- on the other hand, it would be nice to be able to use old setups and not have to change the units just because `gchem` now requires a certain unit.
- if `exf` is used, this the units can be converted already by the `gchem_inscal_*` values.

jm-c commented:
@averdy I am going to merge master-branch into this PR branch, and to do so I will revert the changes to `output.txt` in `global_oce_biogeo_bling/results`  (since it got updated in PR #840 that got merges in since).

jm-c committed:
revert changes/update

jm-c committed:
Merge branch 'master' into gchem_exf

mjlosch commented:
With the merge, the reference results for `global_oce_biogeo_bling` are no longer up to date. As noted earlier, there have been change both here and upstream, so it's a little messy and I didn't even look at the AD part (which is definitely broken and will be fixed later).
@jm-c could you please update the forward reference experiment results so that we can check more easily if further any changes break anything? I could do that, too, but then they would not come from the reference experiements.

jm-c commented:
@mjlosch I could update reference `output.txt`  from `global_oce_biogeo_bling`. but not sure if this would be necessary better:

1. results might change againg before this PR is ready to be merged, and updating this reference output from the refence platform each time "something change" takes time.
2.  It the purpose of the secondary test "exf" is to test the new feature, we may want to keep the primary test unchanged ? If this turns to be an option, it's propably better to keep the current "output.txt" which has been updated in PR #840.
3.  If the reason for result changes from PR #840 also applies to the secondary test, it would make sense to update the secondary test reference output `output.exf.txt`.

Currently the agreement that I get when running testreport is very similar between the 2 tests (primary and secondary "exf"), and is not that great, I agree,

mjlosch commented:
The initialisation of one tracer (Silica) in bling was a bit unorthodox and it was not possible to preserve this in the transition to gchem without excessive exceptions (see discussion above), that's why the results changed in this PR. The changes due to merge were deliberate because of a changed default. The difference that we see now, should be the combination of both changes, if I am not mistaken. But as you can see, I am totally confused now where the changes come from.

We can keep the reference outputs for now, and I'll work with local copies, if I need to.

jm-c commented:
@mjlosch I read more carefully  all the comments you posted earlier regarding update of reference output.txt file, and realized that merging master branch in breaks the complicated way of tracking what modification is responsible for result changes. However, if I understand correctly, the changes from PR #840 were fixing an important aspect of the test (bringing light, was all dark before). For that reason, I thought it would make sense to get this update in.

jm-c commented:
This is not a review but just some thought:

1. it would be easier if there was no update to files that don't really need any for the purpose of this PR, such as, in `pkg/gchem`:
   `gchem_tr_register.F`, `gchem_write_pickup.F`, `GCHEM_OPTIONS.h`, `gchem_forcing_sep.F`, `gchem_output.F` and
 `GCHEM_SIZE.h`    (plus for this one I prefer to keep $#ifdef ALLOW_GCHEM here).
2. I think we need a new switch, like "gchem_useEXF" (or something similar), so that:
   useEXF is for using pkg/exf to get GCM dynamical forcing    -- and --
   gchem_useEXF is for using code from pkg/exf to get BGC forcing.
   Otherwise I would not be able to:
-  use pkg/ebm or pkg/cheapaml with BGC
 - use bulf-formulae and my previous/old dic-forcing settings (data.dic, ...) this case might only be for backward compatibility, but it would be nice to cover it. 
3. I think it's good to have several new header files in `pkg/gchem`, as done here (might even split even more to isolate the exf-interpolation setting so that we limit the number of files where `EXF_INTERP_SIZE.h` needs to be included), and we could always merge some later on if we think it's better.
4. related to above: in `GCHEM_EXF.h` no specific "gchem" or "gchem_' suffix was added to most of these new params. So we need to be sure that this header file will not be included in a BGC source code that does include some previous exf-param setting from before this PR, as this woul results in potential name conflicts. For now, this seems OK to me, but I also don't know much about `pkg/darwin`.
5. leaving units-conversion for later, and looking at names: may be it would be good to start a list of all the fields that each BGC pkg could get from `pkg/gchem` common code. This way we could avoid some confusion. Here I see `gchemSi" with "gchem_silicafile" but in `pkg/dic` we have `silicaSurf` and `silicaDeep` with `DIC_silicaFile` and `DIC_deepSilicaFile`.
6. I did not look at any initialisation issue yet.

jm-c committed:
Merge branch 'master' into gchem_exf

jm-c commented:
To illustrate my previous third point about `EXF_INTERP_SIZE.h`: In `pkg/exf`, the interpolation parameters have been moved to a dedicated header file `EXF_INTERP_PARAM.h`, so that `EXF_INTERP_SIZE.h` does not need to be included everywhere `EXF_PARAM.h` is.
Here, in this current implementation, we will need to include `EXF_INTERP_SIZE.h` each time `GCHEM_EXF.h` is included (seems that it's missing in `gchem_init_fixed.F` and in `bling_init_fixed.F`).

mjlosch committed:
split GCHEM_EXF.h to separate out the interpolation parameters

new file GCHEM_INTERP_PARAM.h, allows to simplify header file inclusion

averdy committed:
new flag gchem_useEXF

averdy committed:
add gchem_ suffix to new exf parameters

averdy committed:
rename new gchem parameters

averdy committed:
update verification expt

mjlosch committed:
fix a typo

mjlosch committed:
first stab at adjusting store directives to new forcing variables

- replace storing at checkpoint levels 2, 3, and 4
- clean up bling-store directives at checkpoint levels 2, 3, and 4
- for some reason that is not yet clear to me , we need to store 3
  gchem fields per forcing (to disk) as opposed to 2 with bling
- adjust MDSIO_BUFF_WH.h in global_oce_biogeo_bling so that it runs
  with the new changes

mjlosch commented:
Oops, now I did commit also the namelist change of the `global_oce_biogeo_bling`-experiment. That was by pure accident. Feel feel to revert that part.

averdy committed:
Fix fields loading when gchem_useEXF=false

averdy commented:
@mjlosch I made progress getting the default experiment to agree with results/output.txt. DIC field still needs work. I'll work on the "exf" case next.

seamanticscience commented:
@jm-c, you mentioned backwards compatibility - should I aim to remove all this stuff?:
```
      NAMELIST /DIC_FORCING/
     &          DIC_windFile, DIC_atmospFile, DIC_silicaFile,
     &          DIC_deepSilicaFile, DIC_iceFile, DIC_parFile,
     &          DIC_chlaFile, DIC_ironFile,
     &          DIC_forcingPeriod, DIC_forcingCycle,
```
since it's relocated to `pkg/gchem`? Or are we trying to maintain `dic_forcing`?

averdy committed:
Load fields in init_varia when gchem_useEXF=FALSE

averdy commented:
With the latest commit, I can reproduce the results for the case gchem_useEXF=FALSE.

mjlosch committed:
minor adjustments

mjlosch committed:
store directive and code adjustment to avoid recomputations

mjlosch commented:
> I did not look at code changes, but would like to ask an open question: Do we plan to have the additional code and forcing arrays defined under a new `pkg/gchem` CPP option ? this might facilitate the transition for, e.g., `pkg/darwin` users ? and in case `pkg/gchem` is used without any BGC purpose, e.g., with `pkg/cfc`, could make the executable "lighter" (smaller memory footprint) and easier to differentiate (with TAF or other AD tools).

That's a good point. I am looking into this (I am using the working title of `GCHEM_ALLOW_FFIELDS` for the CPP-flag) and this is fairly easy to do. I will push something soon.

Since I do not work on this everyday, I keep forgetting the status of the PR: All of the changes in this PR relate to `pkg/bling` only. For `bling` we have moved all code reading forcing fields to `pkg/gchem`, so a CPP flag `GCHEM_ALLOW_FFIELDS` is required (because no forcing code is left in `pkg/bling`). For `pkg/dic`, none of the new functionality is used. One could think of implementing the flag to switch between reading forcing with `pkg/gchem` and `pkg/dic` (i.e. keep the current code in `pkg/dic` and just turn it off when `GCHEM_ALLOW_FFIELDS` is defined). I think @seamanticscience needs to decide that when he tests the code with `dic`. Having said that, we could decide not touch `dic` at all in this PR, but do this in separate one. Again, that's something for the `dic`-users to decide.

mjlosch committed:
Merge branch 'upstream/master' into gchem_exf

mjlosch committed:
for some reason, we do not need these store directives anymore

but we keep them for now as commented-out reminders

mjlosch committed:
introduce GCHEM_ALLOW_FFIELDS to be able to remove a lot of code

- For pkg/cfc we do not need any of the new forcing field code in
  pkg/gchem, so we can reduce the size of the executable
- defined by default in GCHEM_OPTIONS.h

mjlosch committed:
fix previous commit

mjlosch committed:
going around in circles: we do need these store directives

basically reverting commit 08386ec

seamanticscience commented:
@mjlosch I think this P/R should include `pkg/dic`, but I haven't had a decent amount of time to dedicate to it and get the job done. Having a `GCHEM_ALLOW_FFIELDS` flag makes a lot of sense.

seamanticscience committed:
initial work on `pkg/dic` using fields read by `pkg/gchem`

averdy commented:
I was able to match the default results with the input.exf case by adjusting some namelists (in both input and input.exf): 
1) in data, comment out tauThetaClimRelax and tauSaltClimRelax 
2) in data, switch surfQnetFile to surfQFile
3) in data, change periodic forcing to 31-day 
( externForcingPeriod=2678400.,
 externForcingCycle=32140800.,)

mjlosch commented:
> I was able to match the default results with the input.exf case by adjusting some namelists (in both input and input.exf):
> 
> 1. in data, comment out tauThetaClimRelax and tauSaltClimRelax
> 2. in data, switch surfQnetFile to surfQFile
> 3. in data, change periodic forcing to 31-day
>    ( externForcingPeriod=2678400.,
>    externForcingCycle=32140800.,)

Ideally, you only adjust namelists in `input.exf`, e.g. `tauThetaClimRelax` in `data` becomes `climsstTauRelax` in `data.exf`, and the calendar needs to be adjusted to the model calendar, see e.g. `global_ocean_ebm/input.w_exf`.

mjlosch committed:
make physical model's results match ../input with exf

- requires update of output.exf.results
- will be done once we agree on the results of bling

mjlosch committed:
add exf-ice parameters to namelists and initialise

- this was somehow forgotten

mjlosch committed:
adjust namelist for better match between input and input.exf

- still, the input.exf output is different, because the initialisation
of gchemSi is different. In pkg/exf it is initialised to a constant
value and in the default way it is initialised the correctic field at
nIter0. This seems to be necessary, because gchemSi used before
gchem_fields_load is called for the first time. Not sure how to fix this.

mjlosch commented:
With the latest commits, the differences between the input experiment and the input.exf are reduced to the silica field. We I use the output if the input-experiment as a reference I get this (for all fields up to tracer 8)
```
default 10  ----T-----  ----S-----  ----U-----  ----V-----  --PTR 01--  --PTR 02--  --PTR 03--  --PTR 04--  --PTR 05--
G D M    c        m  s        m  s        m  s        m  s        m  s        m  s        m  s        m  s        m  s
e p a R  g  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .
n n k u  2  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d
2 d e n  d  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .

Y Y Y Y 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16  8 16 10> 7<16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 FAIL  global_oce_biogeo_bling.exf
```

The input.exf output is still different, because the initialisation of `gchemSi` is different. With `pkg/exf` it is initialised to a constant value and in the default way it is initialised the correct field at `nIter0`. This seems to be necessary, because `gchemSi `used before `gchem_fields_load` is called for the first time. Not sure how to fix this.

jahn commented:
> The input.exf output is still different, because the initialisation of `gchemSi` is different. With `pkg/exf` it is initialised to a constant value and in the default way it is initialised the correct field at `nIter0`. This seems to be necessary, because `gchemSi `used before `gchem_fields_load` is called for the first time. Not sure how to fix this.

In darwin, the  exf_set_field calls are made early (during the init_varia phase) when necessary, just have to make sure it is not called twice during a timestep.

mjlosch committed:
add a fix for the gchemSi initialisation problem

mjlosch committed:
improve (?) initialisation of gchemIce

- not really necessary, but just for consistency

mjlosch committed:
Merge upstream/master into gchem_exf

mjlosch committed:
This file was not checked in, I do not understand why ...

... and how it was possible for others to even run the
verficiation/global_oce_biogeo_bling.exf experiment

jm-c committed:
Merge branch 'master' into gchem_exf

jm-c commented:
The "USE_QSW" part could be improved (e.g., to work on curvilinear grid) and/ro changed to a run-time switch, but it's independent of this PR purpose, so would be better here to keep it unchanged.

mjlosch committed:
restore USE_QSW in bling_light.F, postpone changes to a future PR

no differences to upstream/main in bling_light.F

averdy commented:
@mjlosch Yes that sounds great

seamanticscience committed:
read gchem forcing into pkg/dic

seamanticscience committed:
Fix unterminated ifdef error and update a few things that got lost somewhere along the way

seamanticscience committed:
trying to fix testreport for `tutorial_global_oce_biogeo` with  GCHEM_ALLOW_FFIELDS undefined

seamanticscience committed:
Revert "trying to fix testreport for `tutorial_global_oce_biogeo` with  GCHEM_ALLOW_FFIELDS undefined"

This reverts commit 96309a35f2db039a2140417be4797b7f4be75507.

seamanticscience committed:
Actually try to fix testreport for `tutorial_global_oce_biogeo` with GCHEM_ALLOW_FFIELDS undefined

seamanticscience committed:
Merge branch 'master' into gchem_exf

seamanticscience committed:
fixed `#undef GCHEM_ALLOW_FFIELDS` in `pkg/dic`

seamanticscience committed:
move bgc forcing from `data.dic` to `data.gchem` and `#define GCHEM_ALLOW_FFIELDS`

seamanticscience commented:
```
run: ./testreport -pass -t tutorial_global_oce_biogeo -devel -of=../tools/build_options/linux_amd64_gfortran -match 14
on : Linux fbd9b8b3873c 6.11.0-1018-azure #18~24.04.1-Ubuntu SMP Sat Jun 28 04:46:03 UTC 2025 x86_64 x86_64 x86_64 GNU/Linux

  OPTFILE=/MITgcm/tools/build_options/linux_amd64_gfortran

default 14  ----T-----  ----S-----  ----U-----  ----V-----  --PTR 01--  --PTR 02--  --PTR 03--  --PTR 04--  --PTR 05--
G D M    c        m  s        m  s        m  s        m  s        m  s        m  s        m  s        m  s        m  s
e p a R  g  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .
n n k u  2  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d
2 d e n  d  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .

Y Y Y Y 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16  7 16  9> 7<16 16 16 16 16 16 16 16 16 16 16 16 12  5  6  6 FAIL  tutorial_global_oce_biogeo
Start time:  Wed Dec 17 02:50:42 UTC 2025
End time:    Wed Dec 17 02:51:47 UTC 2025
```
Hmm, something that affects DIC and O2 would be wind speed...

`pkg/dic` initializes wind in `S/R dic_ini_forcing` by:
```
IF ( DIC_windFile .NE. ' '  ) THEN
        CALL READ_REC_XY_RL( DIC_windFile, wind,
     &                        1, nIter0, myThid )
        _EXCH_XY_RL( wind, myThid )
      ENDIF
```
while `S/R gchem_init_varia` initializes wind by:
```
   IF ( gchem_WindFile .NE. ' ' ) THEN
          CALL READ_REC_XY_RS( gchem_WindFile,gchemWind0,intime0,
     &         nIter0,myThid )
          CALL READ_REC_XY_RS( gchem_WindFile,gchemWind1,intime1,
     &         nIter0,myThid )
        ENDIF
```

mjlosch committed:
add constant pressure to match conversion factor in dic

old default in dic was 1atm, with default Pa2atm = 1013.25e2, so we
need to specify gchem_apres_const = Pa2atm to get exactly 1atm as
initial conditions for forcing before any forcing is read.

mjlosch committed:
fix 2 out of 4 so_box_biogo experiments by adjusting namelists

mjlosch committed:
adjust store directives and namelistes for two AD-verif-exps w/ dic

seamanticscience committed:
update initial "sensible" value for `gchem_apres_const`

seamanticscience committed:
Enable forcing read from `pkg/exf` and add test to `tutorial_global_oce_biogeo`

seamanticscience committed:
Add `STOP` for `gchem_useEXF` if gchem_3DSiFile provided.

seamanticscience committed:
fix indented comment line with `#define EXF_SEAICE_FRACTION`

seamanticscience commented:
Ok, false alarm: for some reason `global_oce_biogeo_bling/results/output.exf.txt` was different to `global_oce_biogeo_bling/results/output.txt`. If I symlink the latter onto the former, both tests pass :)
```
-------------------------------------------------------------------------------
No results email was sent.
Tue Dec 23 08:21:32 EST 2025
run: ./testreport -pass -t global_oce_biogeo_bling -devel -match 14
on : Darwin Jonathans-MBP.lan 23.6.0 Darwin Kernel Version 23.6.0: Thu Dec 19 20:44:10 PST 2024; root:xnu-10063.141.1.703.2~1/RELEASE_ARM64_T6000 arm64

No "OPTFILE" was specified ; genmake2 found and uses:
                         OPTFILE=../../../tools/build_options/darwin_arm64_gfortran
default 14  ----T-----  ----S-----  ----U-----  ----V-----  --PTR 01--  --PTR 02--  --PTR 03--  --PTR 04--  --PTR 05--
G D M    c        m  s        m  s        m  s        m  s        m  s        m  s        m  s        m  s        m  s
e p a R  g  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .
n n k u  2  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d
2 d e n  d  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .

Y Y Y Y 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16>14<16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 pass  global_oce_biogeo_bling
Y Y Y Y 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16>14<16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 pass  global_oce_biogeo_bling.exf
Start time:  Tue Dec 23 08:21:32 EST 2025
End time:    Tue Dec 23 08:27:48 EST 2025
```

seamanticscience committed:
remove additional space in line continuation with `GCHEM_3D_SILICA`

seamanticscience committed:
fix up 3D silicate when `ALLOW_GCHEM_FIELDS` but not exf

seamanticscience committed:
fix multiple definitions of `k` conflicting with `GCHEM_ADD2TR_TENDENCY`

seamanticscience committed:
Add 3d exf files from @jrscott

seamanticscience committed:
swap `3DSi` to `Si3D` and fix a few bugs when `#define GCHEM_3D_SILICA`

seamanticscience committed:
add exf to `so_box_biogeo` compilation

seamanticscience committed:
turn off exf and use gchem to load files

seamanticscience committed:
enable exf in `so_box_biogeo` test `caSat3`

averdy committed:
add adjoint results and update fwd results for expt with gchem_exf

averdy commented:
@seamanticscience The current "bling" results were generated after removing T/S relaxation in data (which changed the physics) but before adding it to data.exf (which reversed the changes). I'm updating the results for the fwd run and adding results for the adjoint.

seamanticscience committed:
first attempt at dic interactive atmosphere co2 with gchem

seamanticscience committed:
Merge branch 'master' into gchem_exf

mjlosch commented on pkg/dic/dic_fields_load.F:
```diff
@@ -47,9 +47,13 @@ SUBROUTINE DIC_FIELDS_LOAD (
+C           only update if dic_int1 = -1 (use gchem fields only in this case)
+C           not if dic_int1 = 3 (interact with atmospheric box)
+            AtmospCO2(i,j,bi,bj)  = gchemapCO2(i,j,bi,bj)
+          ENDIF
```
If you want to avoid IF-statements in a loop, this could easily have its own loop. Similar for the Si3D-code starting at l59

mjlosch commented:
@seamanticscience I was probably wrong about the ALLOW_AUTODIFF parts in `dic_atmos.F`, because the current code basically leads to a nearly empty `s/r dic_atmos_ad` (just resetting some variables to zero). There was "non-zero"/"non-trivial" code in `s/r dic_atmos_ad` at commit [f2b3ff8](https://github.com/MITgcm/MITgcm/pull/876/commits/f2b3ff848a512b517394c3f995dad62578afdc0a) and it's a bit difficult to figure out exactly where this was lost, as some of the intermediate steps (commits) do not compile, but I think it's still there for [cb26c73](https://github.com/MITgcm/MITgcm/pull/876/commits/cb26c73167622182c7e18ada508e29e7e83ccfe3) (after some fixes) but not for [6d91b88](https://github.com/MITgcm/MITgcm/pull/876/commits/6d91b888a60a6f32df8d2ef3c4b72dfa3616aad2). Any idea, what's going on? 
The fact that TAF does not generate anything there means that it thinks that this routine has no effect on the results (for `tutorial_global_oce_biogeo`, `tutorial_dic_adjoffline` alway had an empty `s/r dic_atmos_ad`). Does that make sense?

mjlosch committed:
add new cpp-flag DIC_ALLOW_READ_ATMOSPCO2 (undef by default)

- once this flag is defined, TAF will not generate AD-code for dic_atmos.F

mjlosch committed:
fix recomputations associated with s/r dic_atmos and atmospCO2

remove old hack of avoiding _BEGIN/END_MASTER

mjlosch commented:
@seamanticscience 
after sleeping over it, I am no longer happy with my last night's commit. Instead I will push something, where the `AtmospCO2` is always being set in `dic_atmos.F`. This avoids my stupid new flag (which I remove again in the same commit). In this commit all cases of `dic_int1=-1,0,1,2,3` that are described in the header of the routine are actually handled so that both the reader (that's me) and TAF understand, what's going on in this routine. This means that copying `gchemapco2` to `AtmospCO2` happens here in `dic_atmos.F`, but no longer in `dic_fields_load.F`. If you do not like this solution, then we can revert it, but I think that it is cleaner this way and we do not need to have the complicated logic in `dic_readparms.F` except to stop the code when `dic_int1=-1` without `GCHEM_ALLOW_FFIELDS` defined.

This change may affect results (certainly not in the AD-tests `tutorial_dic_adjoffline` and `tutorial_global_oce_biogeo`, as they seem to be insensitive to any of the changes we have made here), because now the new `AtmopCO2` is only avaialbe after `call dic_atmos` in `s/r gchem_forcing_sep`. This may also cause problems with restarting from pickup as the time levels may not line up. None of this is tested in our current tests.

One could think of moving `call dic_atmos` to the beginning of `gchem_forcing_sep` to avoid this restart issue, but I am not sure what the implications are for the other cases. To me it feels like the natural order of things to first sort out the atmospheric forcing and the run the model (`DIC`). Currently the model (`DIC`) s run and then the forcing for the next timestep is created in `dic_atmos.F`. But I do see that the field `fluxCO2` needs to be initialised properly. 

If you decide that we do not want to go down this route, we can easily revert the single commit.

mjlosch committed:
remove new CPP flag again and reorganize dic_atmos.F

- much simpler to read and understand (and TAF understands it, too)
- restart issue needs to be resolved

mjlosch committed:
disentangle pkg/exf from reading pkg/gchem field with pkg/exf

- makes it possible to use pkg/exf (if compiled) for pkg/gchem fields
  without using pkg/exf for physical fields (with useEXF=F)
- use verification/so_box_biogeo.caSat3 as an example

mjlosch committed:
pkg/cfc now uses gchem forcing fields, adjust 2 verif-experiments

There are no adjoint test cases; also the original cfc_fields_load
uses "sh" instead of "wspeed", but gchem_fields_load uses
"wspeed". Under certain circumstance this can change results. One can
have a local (to pkg/cfc) solution, but it's not clear to me how
important that is. If I use "sh" in gchem_fields_load, I get many new
recomputations, so I leave it for now.

seamanticscience committed:
add prescribed SST and SSS to `pkg/gchem`

seamanticscience committed:
use gchemSST and gchemSSS in `pkg/dic`

seamanticscience committed:
compile fixes

seamanticscience committed:
Make sure dicSST and dicSSS (and model-derived ice fields) are changed after physical model updates. Forcing fields are loaded before physical model updates.

seamanticscience committed:
gchemSST and gchemSSS to `pkg/bling`

jm-c committed:
Merge branch 'master' into gchem_exf

jm-c committed:
fix merge commit from Jan 09

jm-c committed:
remove second copy

remove files that are identical to ../input copy since, by default, all files
from ../input dir are linked to "tr_run.exf" dir

jm-c committed:
back to original version

seamanticscience committed:
remove `kLev` where not needed, replace with surface index "1"

seamanticscience commented on pkg/dic/dic_ini_forcing.F:
```diff
@@ -146,11 +141,7 @@ SUBROUTINE DIC_INI_FORCING( myThid )
-      ELSE
-       kLev = 1
-      ENDIF
+      kLev = 1
```
I see `kLev=1` in `S/R DIC_SURFFORCING_INIT` and `S/R DIC_SURFFORCING`, so i removed them too.

seamanticscience commented:
> I have a fix for the AD problem (add gchem_fields_update.f to `gchem_ad_diff.list` and store the new fields), but I'd rather not push it now, as it is not clear to me if the routine should be added in this PR (it's a new feature/rearranged code, that does change results).

@mjlosch I believe we do need `S/R gchem_forcing_update` (even if we revert `SST` and `SSS` changes) to ensure that up-to-date sea ice fields are passed to the BGC model.

In previous versions, `S/R gchem_fields_load` had [this:](https://github.com/averdy/MITgcm/blob/43472d09a0431ced3fdd70c1ece5156916bf7178/pkg/gchem/gchem_fields_load.F#L491-L518)
```
C------------------------------------------------------------
C If there is no ice file given in data.gchem, try fields from
C pkg/seaice/thsIce

      IF ( gchem_IceFile .EQ. ' ' ) THEN
       DO bj = myByLo(myThid), myByHi(myThid)
        DO bi = myBxLo(myThid), myBxHi(myThid)
#ifdef ALLOW_SEAICE
         IF ( useSEAICE ) THEN
          DO j=1-OLy,sNy+OLy
           DO i=1-OLx,sNx+OLx
            gchemIce(i,j,bi,bj) = AREA(i,j,bi,bj)
           ENDDO
          ENDDO
         ENDIF
#endif
#ifdef ALLOW_THSICE
         IF ( useThSIce ) THEN
          DO j=1-OLy,sNy+OLy
           DO i=1-OLx,sNx+OLx
            gchemIce(i,j,bi,bj) = iceMask(i,j,bi,bj)
           ENDDO
          ENDDO
         ENDIF
#endif
        ENDDO
       ENDDO
      ENDIF
```
which grabs sea ice cover from `pkg/seaice` or `pkg/thsice` at the start of a timestep, _the physical model is updated_, and then `gchemIce` is passed to `pkg/dic` or `pkg/bling`. Thus, the BGC models end up using the initial/previous, rather than the current/updated, sea ice values. In the main repo, `pkg/dic` accesses the current/updated sea ice fraction in `S/R dic_fields_update`; it's just that there's no test with dynamic sea ice and BGC that would have picked up the inadvertently changed results.

@jm-c pointed out that we might need BOTH of these updates to `gchemIce` for `pkg/cfc` to continue working correctly (i.e., it's specifically a problem with BGC models forced "separately").

mjlosch commented:
@seamanticscience I totally agree that this fixes a bug. This bug, however, was there before, just in `pkg/bling`, where the code was inherited from. 

I am just wondering whether we should fix it here, or in a separate PR after we have decided what to do with this PR (to merge or not to merge). Further, an alternative to the new routine could be to `call gchem_fields_load` after the seaice package, e.g. after do_oceanic_phys, just before gchem_calc_tendency, or even just before or at the beginning of  gchem_forcing_sep, when everything should be available.

I can push my fix for the AD, if you need it now, but I'd rather wait until we've decided how to continue with this, see also the latest discussion here: https://github.com/MITgcm/MITgcm/issues/950

seamanticscience committed:
exf 3d interpolation
